From: Ben Hutchings <ben@decadent.org.uk>
Subject: net/sched: Avoid ABI change in 3.10.12

Move psched_ratecfg::linklayer into the padding after the shift field,
and hide it from genksyms.

Make genksyms see tc_ratespec::linklayer as having the previous name
of __reserved.

--- a/include/net/sch_generic.h
+++ b/include/net/sch_generic.h
@@ -682,8 +682,10 @@ struct psched_ratecfg {
 	u64	rate_bps;
 	u32	mult;
 	u16	overhead;
-	u8	linklayer;
 	u8	shift;
+#ifndef __GENKSYMS__
+	u8	linklayer;
+#endif
 };
 
 static inline u64 psched_l2t_ns(const struct psched_ratecfg *r,
--- a/include/uapi/linux/pkt_sched.h
+++ b/include/uapi/linux/pkt_sched.h
@@ -83,7 +83,11 @@ enum tc_link_layer {
 
 struct tc_ratespec {
 	unsigned char	cell_log;
+#if defined(__KERNEL__) && defined(__GENKSYMS__)
+	unsigned char	__reserved;
+#else
 	__u8		linklayer; /* lower 4 bits */
+#endif
 	unsigned short	overhead;
 	short		cell_align;
 	unsigned short	mpu;
